File: src/action/ScriptExecution.php
<?php


namespace Ething\Action;

	
class ScriptExecution extends Action {
	
	
	// attributes validation function
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 'script' => null, 'arguments' => null ),
			$attributes
		);
		
		foreach(array_keys($attributes) as $key){
			
			switch($key){
				
				case 'script':
					
					$scriptId = $attributes[$key];
					
					if(\Ething\ShortId::validate($scriptId)){
						$resource = $ething->get($resourceFilter);
						if(!$resource)
							throw new \Exception("the resource with id '{$scriptId}' does not exist.");
						
						if(!($resource->isTypeof('File') && $resource->mime === 'application/javascript')){
							throw new \Exception("the resource with id '{$scriptId}' is not a script.");
						}
						
					} else throw new \Exception("field '{$key}' must be a script id.");
					
					break;
				
				case 'arguments':
					
					if(!(is_string($attributes[$key]) || is_null($attributes[$key])))
						throw new \Exception("field '{$key}' must be a string or set to null.");
					
					break;
				
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		return true; 
	}
	
	
	protected function call(\Ething\Event\Signal $signal){
		
		$scriptResource = $this->ething()->get($this->script);
		if(!$scriptResource)
			throw new InvalidRuleException("unable to find the script with id '{$scriptResource}'");
		
		$res = \Ething\ScriptEngine::runFromFile($scriptResource, $this->arguments);
		
		if(!$res){
			throw new \Utils\Exception('Unable to execute the script');
		}
		
		if(!$res['ok']){
			throw new \Utils\Exception('Script error');
		}
		
	}
	
}

File: ething/rule/action/ScriptExecution.py
# coding: utf-8
from future.utils import string_types
from .. import Action
from .. import InvalidRuleException
from ething.ScriptEngine import ScriptEngine



class ScriptExecution(Action):
    
    @staticmethod
    def validate (attributes, context):
        
        attributes.setdefault('script', None)
        attributes.setdefault('arguments', None)
        
        for key in attributes:
            
            value = attributes[key]
            
            if key == 'script':
                
                if not( isinstance(value, string_types) and len(value)>0 ):
                    raise Exception("%s: must be a resource id." % key)
                
                script = context['ething'].get(value)
                if not script:
                    raise Exception("%s: no resource found with id = %s." % (key, value))
                
                if not script.isTypeof('File') or script.mime != 'application/javascript':
                    raise Exception("%s: the file %s is not a script" % (key, str(script)))
                
            elif key == 'arguments':
                
                if value is not None:
                    if not( isinstance(value, string_types) and len(value)>0 ):
                        raise Exception("%s: must be a non empty string" % key)
            
            else:
                raise Exception("%s: invalid" % key)
        
        return True
    
    
    
    def call(self, signal):
        
        script = self.ething.get(self['script'])
        if not script:
            raise InvalidRuleException("unable to find the resource with id '%s'" % self['script'])
        
        result = ScriptEngine.runFromFile(script, self['arguments'])
        
        if not result['ok']:
            raise Exception('the script returned with return_code = %d' % result['return_code'] )
        
        
    
    
